File: classes/space.js
import Particle from "./particle.js";
import Square from "./square.js";
import Circle from "./circle.js";
import { getStats } from "./stats.js";

function normalRandom(mu = 0, sigma = 1) {
  let u1 = Math.random();
  let u2 = Math.random();
  
  let z0 = Math.sqrt(-2.0 * Math.log(u1)) * Math.cos(2.0 * Math.PI * u2);
  
  return z0 * sigma + mu;
}
class Space {
  constructor(element, width, height, extraOptions) {
    this.element = element;
    this.numParticlesUpdated = 0;
    const squareWidth = extraOptions?.squareWidth ?? 10;
    let selectorFunction = extraOptions?.selectorFunction;
    let radiusSelectorFunction = extraOptions?.radiusSelectorFunction;
    
    if (!selectorFunction) {
      selectorFunction = () => {
        return { x: Math.random(), y: Math.random() };
      }
    }
    if (!radiusSelectorFunction) {
      radiusSelectorFunction = () => {
        return normalRandom(0, squareWidth);
      }
    }
    
    this.squareHolder = document.createElement('div');
    this.particleHolder = document.createElement('div');

    element.style.width = `${squareWidth * width}px`;
    element.style.height =  `${squareWidth * height}px`;

    element.appendChild(this.squareHolder);
    element.appendChild(this.particleHolder);
    this.setupParticleHolder();
    this.setupSquareHolder();

    this.createCenterOfGravity();

    this.selectorFunction = selectorFunction;
    this.radiusSelectorFunction = radiusSelectorFunction;

    this.width = width;
    this.height = height;
    this.squareWidth = squareWidth;
    this.particles = [];
    this.squares = [];
    for (let i = 0; i < width; i++) {
      this.squares[i] = [];
      for (let j = 0; j < height; j++) {
        const square = new Square(this.particles, { x: i, y: j }, squareWidth);
        this.squareHolder.appendChild(square.elem);
        this.squares[i][j] = square;
      }
    }
  }

  destroy() {
    delete this.squares;
    this.element.innerHTML = '<div class="reset"></div>';
  }

  setupParticleHolder() {
    const elem = this.particleHolder;
    elem.classList.add('particle_holder');
  }

  setupSquareHolder() {
    const elem = this.squareHolder;
    elem.classList.add('square_holder');
  }

  addParticle(x, y) {
    const particle = new Particle(
      { x: x, y: y },
      { x: 0, y: 0 },
      this.particles.length
    );
    particle.updateLog = () => {
      this.numParticlesUpdated++;
    }
    this.particles.push(particle);
    this.particleHolder.appendChild(particle.elem)
    this.updateCenterOfGravity();
  }

  getSquare() {
    let square;
    let count = 20;
    while (!square || count < 0) {
      const selection = this.selectorFunction();
      const x = Math.floor(selection.x * this.width);
      const y = Math.floor(selection.y * this.height);

      square = this.squares[x][y];
      if (square.isUpdating) {
        square = null;
      }
      count--;
    }
    return square;
  }

  update(updateFunc) {
    const square = this.getSquare();
    if (!square) {
      return;
    }

    const callback = () => {
      this.postUpdateNormalize();
    }

    square.update(updateFunc, callback);
  }

  updateGlobal() {
    const forces = this.particles.map(p => {
      return p.getForce(this.particles);
    });
    for (let i in this.particles) {
      const particle = this.particles[i];
      const force = forces[i];
      particle.update(force);
    }
  }

  updateCircle(updateFunc) {
    if (!this.cirlce) {
      this.cirlce = new Circle(
        this.particles, 
      );
    }

    const callback = () => {
      this.postUpdateNormalize();
    }

    const selection = this.selectorFunction();
    const sqW = this.squareWidth;
    const x = selection.x * (this.width + 1) * sqW - (sqW / 2);
    const y = selection.y * (this.height + 1) * sqW - (sqW / 2);

    const r = this.radiusSelectorFunction()

    this.cirlce.setPositionAndRadius(x, y, r);
    this.cirlce.update(updateFunc, callback);
  }

  postUpdateNormalize() {
    const sqW = this.squareWidth;
    for (const particle of this.particles) {
      const pos = particle.position;
      let hitWall = false;

      if (pos.x < sqW / 2) {
        pos.x = sqW / 2;
        hitWall = true;
      } else if (pos.x > (this.width * sqW) - (sqW / 2)) {
        pos.x = (this.width * sqW) - sqW / 2;
        hitWall = true;
      }

      if (pos.y < sqW / 2) {
        pos.y = sqW / 2;
        hitWall = true;
      } else if (pos.y >= (this.height * sqW) - (sqW / 2)) {
        pos.y = (this.height * sqW) - sqW / 2;
        hitWall = true;
      }

      if (hitWall) {
        particle.velocity.x = 0;
        particle.velocity.y = 0;
        particle.updateElemPosition();
      }
    }
    this.updateCenterOfGravity();
  }

  createCenterOfGravity() {
    this.center = document.createElement('div');
    this.center.classList.add('particle', 'centerOfGravity');

    this.particleHolder.appendChild(this.center);
  }

  updateCenterOfGravity() {
    const {x, y} = this.getCenterOfGravity();

    this.center.style.setProperty('--x', x);
    this.center.style.setProperty('--y', y);
  }

  getCenterOfGravity() {
    const xTotal = this.particles.reduce((acc, cur) => {
      return acc + cur.position.x;
    }, 0);
    const yTotal = this.particles.reduce((acc, cur) => {
      return acc + cur.position.y;
    }, 0);

    const x = xTotal / this.particles.length;
    const y = yTotal / this.particles.length;

    return {x: x, y: y}
  }

  centerParticles() {
    const {x, y} = this.getCenterOfGravity();

    const dX = x - (this.width * this.squareWidth) / 2;
    const dY = y - (this.height * this.squareWidth) / 2;

    for (const p of this.particles) {
      p.position.x -= dX;
      p.position.y -= dY;
      p.updateElemPosition();
    }
  }

  calculateStatsForParticles() {
    const particles = this.particles;
    const xValues = particles.map((p) => p.position.x);
    const yValues = particles.map((p) => p.position.y);
  
    return {
      xStats: getStats(xValues),
      yStats: getStats(yValues),
      t: this.numParticlesUpdated,
    };
  }
}

export default Space;

File: threaded/space.py
import random
import math
import threading

from particle import Particle
from circle import Circle

def normal_random(mu=0, sigma=1):
  u1 = random.random()
  u2 = random.random()
  z0 = math.sqrt(-2.0 * math.log(u1)) * math.cos(2.0 * math.pi * u2)
  return z0 * sigma + mu

class Space:
  def __init__(self, width=1200, height=800):
    self.numParticlesUpdated = 0
    self.particles = []
    self.width = width
    self.height = height
    self.xdata, self.distdata, self.devdata = [], [], []

    def radiusSelection():
      sel = -1
      while sel <= 0:
        sel = normal_random(100, 50)
      return sel
    
    self.radiusSelection = radiusSelection

  def addParticle(self, x, y):
    particle = Particle(
      pos={'x': x, 'y': y},
      v={'x': 0, 'y': 0}
    )
    self.particles.append(particle)

  def getCenterOfGravity(self):
    if not self.particles:
      return (0, 0)
    x_total = sum(p.position['x'] for p in self.particles)
    y_total = sum(p.position['y'] for p in self.particles)
    x = x_total / len(self.particles)
    y = y_total / len(self.particles)
    return (x, y)

  def centerParticles(self):
    x, y = self.getCenterOfGravity()
    dX = x - (self.width / 2)
    dY = y - (self.height / 2)
    for p in self.particles:
      p.position['x'] -= dX
      p.position['y'] -= dY

  def draw(self, batch, circleFunction):
    self.drawables = []
    for particle in self.particles:
      self.drawables.append(particle.draw(batch, circleFunction))

  def update_global(self):
    forces = [p.getForce(self.particles) for p in self.particles]

    for i, particle in enumerate(self.particles):
      force = forces[i]
      particle.update(force)

  def update_circle(self):
    circle = Circle(self.particles)

    x = random.uniform(0, self.width)
    y = random.uniform(0, self.height)

    r = self.radiusSelection()
    circle.setPositionAndRadius(x, y, r)

    particles_updated = circle.update()
    avgerage_distance = self.average_distance_from_center()
    std_dev = self.calculate_stddev(sample=True)
    avg_std = (std_dev['std_x'] + std_dev['std_y']) / 2
    with threading.Lock():
      self.numParticlesUpdated += particles_updated
      self.xdata.append(self.numParticlesUpdated)
      # std = self.calculate_stddev(sample=True)
      self.distdata.append(avgerage_distance)
      self.devdata.append(avg_std)

  def calculate_stddev(self, sample=False):
    particles = self.particles
    if not particles:
      return {"std_x": 0.0, "std_y": 0.0}
    
    n = len(particles)
    mean_x = sum(p.position['x'] for p in particles) / n
    mean_y = sum(p.position['y'] for p in particles) / n

    var_x = sum((p.position['x'] - mean_x) ** 2 for p in particles)
    var_y = sum((p.position['y'] - mean_y) ** 2 for p in particles)

    # Population vs Sample
    denom = n - 1 if sample and n > 1 else n

    std_x = math.sqrt(var_x / denom)
    std_y = math.sqrt(var_y / denom)

    return {"std_x": std_x, "std_y": std_y}

  def average_distance_from_center(self):
    if not self.particles:
      return 0.0

    cx, cy = self.getCenterOfGravity()
    total_distance = 0.0

    for p in self.particles:
      dx = p.position['x'] - cx
      dy = p.position['y'] - cy
      total_distance += math.sqrt(dx*dx + dy*dy)

    return total_distance / len(self.particles)